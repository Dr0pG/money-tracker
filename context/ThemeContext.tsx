import AsyncStorage from "@react-native-async-storage/async-storage";
import React, {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from "react";

// Define the Theme type
export type Theme = "light" | "dark";

// Define the context type
interface ThemeContextType {
  theme: Theme;
  toggleTheme: () => void;
}

// Create the context
const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

interface ThemeProviderProps {
  children: ReactNode;
  initialTheme?: Theme;
}

export const ThemeProvider: React.FC<ThemeProviderProps> = ({
  children,
  initialTheme = "dark",
}) => {
  const [theme, setTheme] = useState<Theme>(initialTheme);

  useEffect(() => {
    const loadTheme = async () => {
      const storedTheme = await AsyncStorage.getItem("theme");
      if (storedTheme === "light" || storedTheme === "dark") {
        setTheme(storedTheme);
      }
    };
    loadTheme();
  }, []);

  const toggleTheme = async () => {
    const newTheme: Theme = theme === "light" ? "dark" : "light";
    setTheme(newTheme);
    await AsyncStorage.setItem("theme", newTheme);
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useTheme = (): ThemeContextType => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error("useTheme must be used within a ThemeProvider");
  }
  return context;
};
