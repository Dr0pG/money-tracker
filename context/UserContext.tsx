import React, { createContext, useContext, ReactNode } from "react";
import auth, { FirebaseAuthTypes } from "@react-native-firebase/auth";

// Define the context type
interface UserContextType {
  user: FirebaseAuthTypes.User | null;
}

// Create the context
const UserContext = createContext<UserContextType | undefined>(undefined);

interface UserProviderProps {
  children: ReactNode;
}

export const UserProvider = ({ children }: UserProviderProps) => {
  const currentUser: FirebaseAuthTypes.User | null = auth().currentUser;

  return (
    <UserContext.Provider value={{ user: currentUser }}>
      {children}
    </UserContext.Provider>
  );
};

export const getUser = (): UserContextType => {
  const context = useContext(UserContext);
  if (!context) {
    throw new Error("Can't get any user!");
  }
  return context;
};
