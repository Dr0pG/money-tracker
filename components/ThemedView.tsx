import React, { memo, useEffect } from "react";
import { type ViewProps } from "react-native";
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
} from "react-native-reanimated";
import { useTheme } from "@/context/ThemeContext";
import { useThemeColor } from "@/hooks/useThemeColor";
import Durations from "@/constants/Durations";

export type ThemedViewProps = ViewProps & {
  lightColor?: string;
  darkColor?: string;
};

const ThemedView = ({
  style,
  lightColor,
  darkColor,
  ...otherProps
}: ThemedViewProps) => {
  const { theme } = useTheme();
  const lightBg = lightColor || useThemeColor({}, "background");
  const darkBg = darkColor || useThemeColor({}, "background");

  const animatedColor = useSharedValue(theme === "light" ? lightBg : darkBg);

  useEffect(() => {
    animatedColor.value = withTiming(theme === "light" ? lightBg : darkBg, {
      duration: Durations.colorChanged,
    });
  }, [theme, lightBg, darkBg]);

  const animatedStyle = useAnimatedStyle(() => ({
    backgroundColor: animatedColor.value,
  }));

  return <Animated.View style={[animatedStyle, style]} {...otherProps} />;
};

export default memo(ThemedView);
