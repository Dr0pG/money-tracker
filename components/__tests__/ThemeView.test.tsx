import React from "react";
import { render } from "@testing-library/react-native";
import ThemedView from "@/components/ThemedView";
import { ThemeProvider } from "@/context/ThemeContext";

describe("ThemedView", () => {
  it("renders correctly with light theme", () => {
    const { toJSON } = render(
      <ThemeProvider initialTheme="light">
        <ThemedView testID="themed-view" />
      </ThemeProvider>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it("renders correctly with dark theme", () => {
    const { toJSON } = render(
      <ThemeProvider initialTheme="dark">
        <ThemedView testID="themed-view" />
      </ThemeProvider>
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
