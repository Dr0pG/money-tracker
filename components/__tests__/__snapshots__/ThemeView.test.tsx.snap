// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemedView renders correctly with dark theme 1`] = `
<View
  collapsable={false}
  jestAnimatedProps={
    {
      "value": {},
    }
  }
  jestAnimatedStyle={
    {
      "value": {
        "backgroundColor": "#1A181B",
      },
    }
  }
  jestInlineStyle={
    [
      undefined,
    ]
  }
  style={
    [
      {
        "backgroundColor": "#1A181B",
      },
      undefined,
    ]
  }
  testID="themed-view"
/>
`;

exports[`ThemedView renders correctly with light theme 1`] = `
<View
  collapsable={false}
  jestAnimatedProps={
    {
      "value": {},
    }
  }
  jestAnimatedStyle={
    {
      "value": {
        "backgroundColor": "#E2E2E8",
      },
    }
  }
  jestInlineStyle={
    [
      undefined,
    ]
  }
  style={
    [
      {
        "backgroundColor": "#E2E2E8",
      },
      undefined,
    ]
  }
  testID="themed-view"
/>
`;
